brainstorming.md

```diff
@@ -0,0 +1,508 @@
+
+The composed types are:
+
+- SOMAExperiment - a specialization and extension of SOMACollection, codifying a set of naming and indexing conventions to represent annotated, 2-D matrix of observations across _multiple_ sets of variables.
+
+In this document, the term `dataframe` implies something akin to an Arrow `RecordBatch`, R `data.frame` or Python `pandas.DataFrame`, where:
+
+- multiple columns may exist, each with a string column name
+- all columns are individually typed, monomorphic, and contain simple types (eg, int64)
+- all columns are of equal length
+- one or more columns may be indexed
+
+All SOMA data objects are named with URIs.
+
+## Base Type System
+
+The SOMA API borrows its base type system from the Arrow language-agnostic in-memory system for data typing and serialization ([format](https://arrow.apache.org/docs/format/Columnar.html)). The SOMA API is intended to be used with an Arrow implementation such as [PyArrow](https://arrow.apache.org/docs/python/) or the [Arrow R package](https://arrow.apache.org/docs/r/), or other libraries which interoperate with Arrow (eg, Pandas).
+
+Where SOMA requires an explicit typing system, it utilizes the Arrow types and schema. SOMA has no specific requirements on the type or serialization system used by the underlying storage engine, other than it be capable of understanding and representing the Arrow types. It is expected that any given implementation of SOMA will have limits on the underlying capabilities of its data type system (eg, just because you can express a type in he Arrow type system does not mean all SOMA implementations will understand it).
+
+### Type definitions used in this document
+
+In the following doc:
+
+- `primitive` types in this specification refer to Arrow primitive types, eg, `int32`, `float`, etc.
+- `string` refers to Arrow UTF-8 variable-length `string`, ie, `List<Char>`.
+- `simple` types include all primitive types, plus `string`.
+
+Other Arrow types are explicitly noted as such, eg, `Arrow RecordBatch`.
+
+> ⚠️ **Issue** - are there parts of the Arrow type system that we wish to _explicitly exclude_ from SOMA? I have left this issue open (ie, no specific text) for now, thinking that we can come back and subset as we understand what complex types are required, and how much flexibility should be in this spec. We clearly need some complex types (eg, RecordBatch, List, etc) as they are implied by `string`, etc. My own preference would be to mandate a small set of primitive types, and leave the rest open to implementations to support as they feel useful.
+
+> ⚠️ **Issue** - the above uses Arrow `string`. Should we be using `large_string` instead (no 2GB cap)?
+
+## Metadata
+
+All SOMA objects may be annotated with a small amount of metadata:
+
+- `metadata` (map[string, simple]) - user metadata, specified as string-keyed mapping to any [**simple**](#type-definitions-used-in-this-document) type (eg, int32, string, etc).
```

Comment:
i know anndata often has metadata values which are lists of strings -- so defining "simple" to exclude "list of string" is, i believe, going to block the the ingest-anndata use-case immediately -- this known, concrete use-case warrants the complexity of supporting list of string in metadata

as a workaround, maybe map list of strings `["a","b","c"]` from other storages into comma-joined `"a,b,c"` -- ? but i prefer including support for list-of-string in our bare minimum

Reply:
> Besides "short list of simple types", what else do you recommend as a "minimum" requirement?

@bkmartinjr my minimum list includes your simples, and list thereof

Reply:
Not all lists can be expressed as an Arrow array or table. For example, I believe that they need to be monomorphic.

My suggestion:  anything that can be stored as an Arrow Array is "OK".  Indicate that they should be "small", and leave the definition of "small" up to the implementation.

Thoughts?

Reply:
i am happy with us supporting arrays of ints and arrays of strings in metadata as i know that already in my data-explorations i've seen lots of both in the wild -- we should support these from day one

Reply:
I have added Arrow arrays to the list of types supported by metadata (values).  Keys are still strictly strings.

Reply:
I can think of lots of examples where complex metadata _exist_ (lists or arrays of strings, json blobs) but none where complex metadata are _required_. 

> i know anndata often has metadata values which are lists of strings -- so defining "simple" to exclude "list of string" is, i believe, going to block the the ingest-anndata use-case immediately

I think this is the right problem to focus on. Because AnnData is more permissive, we're often going to either (a) fail ingestion or (b) introduce data loss. I favor loss of data + a warning during ingestion when that occurs is a good approach given the cases I'm aware of.

Reply:
In offline discussion of this, it was also pointed out that the SOMACollection is the basis of all composed types, and it can store arbitrary dataframe/array data with a user-specified name. This is a relatively full-featured substitute for complex metadata, and provides the affordances of those more advanced types (eg, incremental read/write).  Decision taken to revert the metadata back to simple/primitive types.

Reply:
> This is a relatively full-featured substitute for complex metadata,

This is basically where AnnData has landed for now too.

However we're going to expand the per element metadata within anndata. This is needed for a number of use cases around spatial data, and could be quite useful for handling provenance. Right now, I'm thinking this metadata must be json serializable.

Reply:
> we're going to expand the per element metadata within anndata

Can you elaborate on what way you will expand it?  Do you mean adding a `uns` equivalent to more objects (for example, `adata.obsm.uns['foo'] = 'bar'`?  Or that the supported type system for `uns` values will expand?  Or something else?

Really appreciate your engagement!

Reply:
Many thanks for the concrete example.

This example (JSON serialized string on an obsm array) fits our current design thinking (string/primitive metadata values on all array/dataframe objects).